libraries/views.py

```diff
@@ -194,25 +200,41 @@ class LibraryListByCategory(LibraryList):
 
     def get_context_data(self, **kwargs):
         context = super().get_context_data(**kwargs)
-        context["library_list"] = self.get_results_by_category(
+        context["library_versions_by_category"] = self.get_results_by_category(
             version=context.get("version")
         )
         return context
 
     def get_results_by_category(self, version: Version | None):
-        queryset = super().get_queryset()
-        filter_kwargs = {"libraries__versions__name": version} if version else {}
+        # Define filter kwargs based on whether version is provided
+        version_filter = {"version__name": version} if version else {}
+        category_filter = {"libraries__versions__name": version} if version else {}
+
+        library_versions_qs = LibraryVersion.objects.filter(**version_filter)
         categories = (
-            Category.objects.filter(**filter_kwargs).distinct().order_by("name")
-        )
-        categories = categories.prefetch_related(
-            Prefetch("libraries", queryset=queryset, to_attr="prefetched_libraries")
+            Category.objects.filter(**category_filter)
+            .distinct()
+            .order_by("name")
+            .prefetch_related(
+                Prefetch(
+                    "libraries__library_version",
+                    queryset=library_versions_qs,
+                    to_attr="prefetched_library_versions",
+                )
+            )
         )
-        results_by_category = [
-            {"category": category, "libraries": category.prefetched_libraries}
+
+        return [
+            {
+                "category": category,
+                "library_version_list": [
+                    lv
+                    for library in category.libraries.all()
+                    for lv in getattr(library, "prefetched_library_versions", [])
+                ],
+            }
             for category in categories
         ]
```

Comment:
IMO nesting comprehensions three deep is too much. It's a bit hard for me to parse what's going on.

See Google's style guide for Python for [an argument in this vein](https://google.github.io/styleguide/pyguide.html#274-decision).

Reply:
Yeah, fair point, I'm not a huge fan of those either generally. I've now committed a version that should be easier to read.